src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Folding/FoldingRangeEndpoint.cs

```diff
@@ -136,7 +140,35 @@ public FoldingRangeRegistrationOptions GetRegistrationOptions(FoldingRangeCapabi
                 mappedRanges.AddRange(ranges);
             }
 
-            return new Container<FoldingRange>(mappedRanges);
+            var finalRanges = mappedRanges.Select(r => FixFoldingRangeStart(r, codeDocument)).ToImmutableArray();
+            return new Container<FoldingRange>(finalRanges);
+        }
+
+        /// <summary>
+        /// Fixes the start of a range so that the offset of the first line is the last character on that line. This makes
+        /// it so collapsing will still show the text instead of just "..."
+        /// </summary>
+        private FoldingRange FixFoldingRangeStart(FoldingRange range, RazorCodeDocument codeDocument)
+        {
+            Debug.Assert(range.StartLine < range.EndLine);
```

Comment:
Can we do this because we throw out `range.StartLine can == range.EndLine` at another layer prior to this method?

Reply:
We don't, but folding ranges _shouldnt_ report for a single line. That said, we could filter those out explicitly instead of adding this debug assert. If they do occur it's an error from something else

Reply:
lol speak of the devil:


```
---------------------------
Assertion Failed: Abort=Quit, Retry=Debug, Ignore=Continue
---------------------------


   at Microsoft.AspNetCore.Razor.LanguageServer.Folding.FoldingRangeEndpoint.FixRangeStart(Range mappedRange, RazorCodeDocument codeDocument)
   at Microsoft.AspNetCore.Razor.LanguageServer.Folding.FoldingRangeEndpoint.<HandleCoreAsync>d__10.MoveNext()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c__DisplayClass4_.<OutputAsyncCausalityEvents>b__0()
   at System.Runtime.CompilerServices.TaskAwaiter.<>c__DisplayClass11_0.<OutputWaitEtwEvents>b__0()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.SetResult(TResult result)
   at OmniSharp.Extensions.JsonRpc.ResponseRouter.ResponseRouterReturnsImpl.<Returning>d__4`1.MoveNext()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.MoveNextRunner.Run()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c__DisplayClass4_0.<OutputAsyncCausalityEvents>b__0()
   at System.Runtime.CompilerServices.TaskAwaiter.<>c__DisplayClass11_0.<OutputWaitEtwEvents>b__0()
   at System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(Action action, Boolean allowInlining, Task& currentTask)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task`1.TrySetResult(TResult result)
```

Does this mean either C# or HTML is being a bad citizen?

Reply:
We should probably be smart and ignore but also we should file bugs on the C# / HTML side for returning 0 line difference folding ranges too given there will still be computation spent serializing/desrializing all of them

Reply:
> but folding ranges shouldnt report for a single line

FWIW, we have [a work item](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1288779) on the HTML backlog to provide folding ranges for base64 strings as a parity feature with WebEssentials, which would typically be on a single line.  (No idea when we'll get to it, but just pointing out that it could come up).